### PR TITLE
Clarify event id vs. "at-least-once" delivery

### PR DESCRIPTION
**Pull request by a1exsh**
This came up again recently and it seems that the relation should be further elaborated.
Random UUIDs are not suitable in certain situations (without intermediate persistence),
even though they are guaranteed to be "unique from the perspective of the owning application",
as required in rule 211.

### CODE REVIEW

File: chapters/events.adoc [hunk 1]
```diff
@@ -855,12 +855,15 @@ events within a given event type's stream must have unique identifiers.
 This allows consumers to process the `eid` to assert the event is unique
 and use it as an idempotency check.
 
-Note that uniqueness checking of the `eid` might be not enforced by
-systems consuming events and it is the responsibility of the producer to
-ensure event identifiers do in fact distinctly identify events. A
-straightforward way to create a unique identifier for an event is to
-generate a UUID value.
+It is the responsibility of the producer to ensure event identifiers do
+in fact distinctly identify events. A straightforward way to create a
```
Comment (tfrauenstein) [inline chapters/events.adoc:859]: pls. add:  
in fact distinctly identify events. --> in fact distinctly identify events published for a specific event type.

File: chapters/events.adoc [hunk 2]
```diff
@@ -855,12 +855,15 @@ events within a given event type's stream must have unique identifiers.
...
+unique identifier for an event is to generate a UUID value. However, due
+to "at-least-once" delivery guarantees that most message brokers provide,
+event producers need to ensure that subsequent attempts to publish an
+event do use the same identifier as the initial (possibly failed) attempt.
 
+Considerations for event consumers are further detailed in <<212>> and
+<<214>>.
```
Comment (tfrauenstein) [inline chapters/events.adoc:867]: Propose to change this sentence to: 
Event identifiers facilitate event duplicate detection by event consumers -- see <<214>>.

**Review by tfrauenstein — COMMENTED** [review]
Thx. looks good -- just two minor change proposals.

**Comment by a1exsh** [comment]
@tfrauenstein thank you for your comments!

Do you think it is appropriate here to touch upon the two common strategies for publishers to ensure stable event id on retry, namely: intermediate persistence and consistent hashing?

**Comment by tfrauenstein** [comment]
Yes, adding a 'design hint' might be helpful here, if it is not too long. 
But let us clarify it offline to avoid ping-pong here.

File: chapters/events.adoc [hunk 3]
```diff
@@ -855,12 +855,16 @@ events within a given event type's stream must have unique identifiers.
...
 
-Note that uniqueness checking of the `eid` might be not enforced by
-systems consuming events and it is the responsibility of the producer to
-ensure event identifiers do in fact distinctly identify events. A
-straightforward way to create a unique identifier for an event is to
-generate a UUID value.
+It is the responsibility of the producer to ensure event identifiers do
+in fact distinctly identify events published for a specific event type.
+A straightforward way to create a unique identifier for an event is to
+generate a UUID value. However, due to "at-least-once" delivery guarantees
```
Comment (tfrauenstein) [inline chapters/events.adoc:861]: As discussed, I propose to update this last sentence by:

However, event producers need to ensure that retries to publish an event (e.g. as 
reaction to temporary Nakadi call failures), use the same event identifier as the 
initial (possibly failed) attempt.

Hint: Using the same eid for retries can be ensured e.g. by deterministic eid 
compute functions only based on event inputs (uuid without random ingredients)  
or some form of intermediate persistency like dead letter queues.

**Comment by tfrauenstein** [comment]
👍

**Comment by a1exsh** [comment]
@tkrop @tfrauenstein please have another look

**Comment by tfrauenstein** [comment]
👍

**Comment by tkrop** [comment]
:+1:

**Pull request merged**